paasta_tools/autoscaling/autoscaling_cluster_lib.py

```diff
@@ -234,53 +255,54 @@ def wait_and_terminate(self, slave, drain_timeout, dry_run, region=None, should_
-                            time.sleep(300)
-                        self.log.info("TERMINATING: {} (Hostname = {}, IP = {})".format(
-                            instance_id,
-                            slave.hostname,
-                            slave.ip,
-                        ))
-                        try:
-                            ec2_client.terminate_instances(InstanceIds=[instance_id], DryRun=dry_run)
-                        except ClientError as e:
-                            if e.response['Error'].get('Code') == 'DryRunOperation':
-                                pass
-                            else:
-                                raise
-                        break
-                    else:
-                        self.log.info("Instance {}: NOT ready to kill".format(instance_id))
-                    self.log.debug("Waiting 5 seconds and then checking again")
-                    time.sleep(5)
+            while True:
+                instance_id = slave.instance_id
+                if not instance_id:
+                    self.log.warning(
+                        "Didn't find instance ID for slave: {}. Skipping terminating".format(slave.pid),
+                    )
+                    continue
+                # Check if no tasks are running or we have reached the maintenance window
+                if self.can_kill(slave.hostname, should_drain, dry_run, timer):
+                    self.log.info("TERMINATING: {} (Hostname = {}, IP = {})".format(
+                        instance_id,
+                        slave.hostname,
+                        slave.ip,
+                    ))
+                    try:
+                        ec2_client.terminate_instances(InstanceIds=[instance_id], DryRun=dry_run)
+                    except ClientError as e:
+                        if e.response['Error'].get('Code') == 'DryRunOperation':
+                            pass
+                        else:
+                            raise
+                    break
+                else:
+                    self.log.info("Instance {}: NOT ready to kill".format(instance_id))
+                self.log.debug("Waiting 5 seconds and then checking again")
+                await asyncio.sleep(5)
         except TimeoutError:
             self.log.error("Timed out after {} waiting to drain {}, now terminating anyway".format(
                 drain_timeout,
                 slave.pid,
             ))
             try:
-                ec2_client.terminate_instances(InstanceIds=instance_id, DryRun=dry_run)
+                ec2_client.terminate_instances(InstanceIds=[instance_id], DryRun=dry_run)
             except ClientError as e:
                 if e.response['Error'].get('Code') == 'DryRunOperation':
                     pass
                 else:
                     raise
 
-    def scale_resource(self, current_capacity, target_capacity):
+    async def scale_resource(self, current_capacity, target_capacity):
```

Comment:
This function needs an `await` call for the branches other than `delta < 0`

Reply:
Why?

If it really does need that, then can you say so in the docstring?

Reply:
Python will throw a `RuntimeWarning` if an `async` function doesn't `await` on anything.  I was running into issues with this on testing, but it's probably worth adding the proper flag to not raise the RuntimeWarnings as Errors (and then I can remove all the extra `asyncio.sleep(0.1)`).